### DISCUSSION TITLE
Logging

### DISCUSSION THREAD

**Discussion by ghost** [eModbus/eModbus#342] [Q&A] [answered]
I am using the RTU06Example almost verbatim in an ESP32 with Platformio  and it works beautifully. The ini library is eModbus@^1.7.1. The app sends a READ_HOLD_REGISTER request to a sensor and prints the reply.
However, as soon as I remove all references to logging (including the header), it records sending the request and then hangs. I have tried just including the Logging.h header and fiddling with LOG_LEVEL must this seems to have no effect. Am I missing something here?

**Comment by Miq1** [comment]
Never tried to remove all logging... Since the log lines all are preprocessor macros that will vanish if the log level is lower than their required, a ``-DLOG_LEVEL=0`` should you get rid of all these lines.
However, the logging inside the library code is assumed to be present at places where the library has no solution for something happening else than shouting it out, so may be the program structure can degenerate if no log line is compiled, like an ''if'' clause with no log line as a body may grab the next code line instead. I doubt that, though, but you never know...
I will check for that and keep you posted.

**Reply by Miq1** [reply]
On a second thought, if you are using the example as is, it will of course be completely silent with logging disabled, since all output is done by logging lines. 
Would you be able to post your complete sketch here?

**Reply by ghost** [reply]
The app sends 3 modbus requests to a sensor. While not my normal code I have simulated the sequence in the **loop** function. It waits before sending a new request until the previous is acknowledged. Because the sensor will go to sleep after 5 seconds without a request, a few characters are sent to wake it up (ln 156). The sensor ignores these characters.
As you see it below it all works. However, on my system if line 50 (the LOG_N) is commented the code hangs after sending the second request.
I would be very interested in your comments.
Regards, Ron
```
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================

// Example code to show the usage of the eModbus library. 
// Please refer to root/Readme.md for a full description.

// Includes: <Arduino.h> for Serial etc.
#include <Arduino.h>

// Include the header for the ModbusClient RTU style
#include "ModbusClientRTU.h"

// For demonstration, use the LOG statements for output
#define LOG_LEVEL 0
#define LOCAL_LOG_LEVEL 0
#include "Logging.h"

// Create a ModbusRTU client instance
// In my case the RS485 module had auto halfduplex, so no parameter with the DE/RE pin is required!
ModbusClientRTU MB;
uint32_t old;
uint32_t Token = 0;

uint32_t Token_1;
uint32_t Token_2;
uint32_t Token_3;

bool Processed;
uint16_t Count;

//-------------------------------------------------------------------
// Define an onData handler function to receive the regular responses
// Arguments are Modbus server ID, the function code requested, the 
// message data and length of it, plus a user-supplied token to 
// identify the causing request
//
void handleData(ModbusMessage response, uint32_t token) 
{
   Serial.printf("\nHandleData %d %d %d %d\n", token, Token_1, Token_2, Token_3);

   // Only print out result of the "real" example - not the request preparing the field
   if (token > 0) 
   {
      float value;
      uint16_t rep;
      uint16_t v;
      
      LOG_N("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
//      Serial.printf("\nResponse: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
      if (token == Token_1)
      {
         rep = response.get(3, value);
         Serial.printf("Pressure %f\n", value);
         Processed = true;
      }

      if (token == Token_2)
      {
         rep = response.get(3, value);
         Serial.printf("Temp %f\n", value);
         Processed = true;
      }

      if (token == Token_3)
      {
         rep = response.get(3, value);
         Serial.printf("Level %f\n", value);
         Processed = true;
      }
   }
}

//-------------------------------------------------------------------
// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  LOG_E("Error response: %02X - %s\n", (int)me, (const char *)me);
}

//-------------------------------------------------------------------
// Setup() - initialization happens here
//
void setup() 
{
   // Init Serial monitor
   Serial.begin(115200);
   while (!Serial) {}
   Serial.println("__ OK __");

   // Set up Serial2 connected to Modbus RTU (Fill in your data here!)
   //
   RTUutils::prepareHardwareSerial(Serial2);
   Serial2.begin(19200, SERIAL_8E1, GPIO_NUM_16, GPIO_NUM_17);

   MB.useModbusRTU();
   
   // Set up ModbusRTU client.
   // - provide onData handler function
   //
   MB.onDataHandler(&handleData);

   // - provide onError handler function
   //
   MB.onErrorHandler(&handleError);

   // Set message timeout to 2000ms
   //
   MB.setTimeout(2000);

   // Start ModbusRTU background task
   //
   MB.begin(Serial2);

   // The Token value is used in handleData to synch the reply with
   // the sent packet.avoid the output for 
   //
   Token = 10;

   old = millis();

   Processed = true;
   Token_1 = 99;
   Token_2 = 99;
   Token_3 = 99;
}

uint32_t theDelay = 2000;

//-------------------------------------------------------------------
// loop() - nothing done here today!
//
void loop() 
{
   Error err;

   if (millis() > old + theDelay)
   {
      if (Processed)
      {
         Processed = false;

         Serial.printf("\nCount %d\n", Count);

         if (Count == 0)
         {
// For this demo this is only used once at startup as the inter-group delay
// is less 5 secs (see ln 200).
            Serial.println("Wakeup");
            // The sensor needs at least a byte to wake up if it has
            // been un-accessed for more than 5 secs.
            Serial2.println();
            theDelay = 100;
            Processed = true;
         }

         if (Count == 1)
         {
            Serial.printf("\nPressure req %d\n", Token);
            err = MB.addRequest(Token, 1, READ_HOLD_REGISTER, 0x0025, 4);
            Token_1 = Token++;
            if (err!=SUCCESS) 
            {
               ModbusError e(err);
               LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
            }
            theDelay = 20;
         }

         if (Count == 2)
         {
            Serial.printf("Temp req %d\n", Token);
            err = MB.addRequest(Token, 1, READ_HOLD_REGISTER, 0x002d, 4);
            Token_2 = Token++;
            if (err!=SUCCESS) 
            {
               ModbusError e(err);
               LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
            }
            theDelay = 20;
         }

         if (Count == 3)
         {
            Serial.printf("Level req %d\n", Token);
            err = MB.addRequest(Token, 1, READ_HOLD_REGISTER, 0x0035, 4);
            Token_3 = Token++;
            if (err!=SUCCESS) 
            {
               ModbusError e(err);
               LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
            }
            theDelay = 3000;
         }

         if (++Count > 3) Count = 1;
      }
      old = millis();
   }
}

```

**Reply by Miq1** [reply]
Can you add a sample set of lines of the output, please? 

A few minor things popped up - nothing relevant, though:
- Count is never initialized, you are relying on memory being zero on boot.
- the check for millis() in loop() will fail eventually when the uint32_t value will wrap around. 
- all requests are serialized, so no need to loop around for 20ms between your requests. Just do three addRequest() calls in a row.
- (this was wrong)

**Comment by Miq1** [comment]
Duplicate post removed.

**Comment by ghost** [comment]
Hi Michael

Thanks for your comments. Here are my answers:

1.	In C and C++ all globals are zeroed before calling main.
2.	I did mention this code was just to be as simple as possible. 
3.	That was my initial understanding, but I had separated the requests as trying to understand what was wrong. As the initial “wakeup” character is not a modbus packet that remains separate.
4.	This was also trying to isolate anything in my code from the library. The code below shows the current example. 

 

Note I have added a yield() call at the end (I am using FreeRTOS elsewhere) but this had no effect on the library problem.

 

Regards,
Ron
```
void loop() 
{
   Error err;

   if (millis() > old + theDelay)
   {
      Serial.printf("\nCount %d\n", Count);
      if (Count == 0)
      {
         Serial.println("Wakeup");
         // The sensor needs at least a byte and 20mSecs to wake up if it has
         // been un-accessed for more than 5 secs.
         //
         Serial2.println();
         theDelay = 20;
      }
      if (Count == 1)
      {
         Serial.printf("\nPressure req %d\n", Token);
         err = MB.addRequest(Token, 1, READ_HOLD_REGISTER, 0x0025, 4);
         Token_1 = Token++;
         if (err!=SUCCESS) 
         {
            ModbusError e(err);
            LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
         }
         Serial.printf("Temp req %d\n", Token);
         err = MB.addRequest(Token, 1, READ_HOLD_REGISTER, 0x002d, 4);
         Token_2 = Token++;
         if (err!=SUCCESS) 
         {
            ModbusError e(err);
            LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
         }
         Serial.printf("Level req %d\n", Token);
         err = MB.addRequest(Token, 1, READ_HOLD_REGISTER, 0x0035, 4);
         Token_3 = Token++;
         if (err!=SUCCESS) 
         {
            ModbusError e(err);
            LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
         }
         theDelay = 10000;
      }
      if (++Count > 1) Count = 0;
      old = millis();
   }
   yield();
}
```

**Comment by ghost** [comment]
Here is the output when the LOG_N at line 50 is included. This is what I expected:

 

mode:DIO, clock div:2    

load:0x3fff0030,len:1184 

load:0x40078000,len:13232

load:0x40080400,len:3028 

entry 0x400805e4

__ OK __

 

Count 0

Wakeup 

 

Count 1        

 

Pressure req 10

Temp req 11    

Level req 12   

 

HandleData 10 10 11 12

[N] 2197| main.cpp             [  50] handleData: Response: serverID=1, FC=3, Token=0000000A, length=11:

Pressure 0.003629

 

HandleData 11 10 11 12

[N] 2333| main.cpp             [  50] handleData: Response: serverID=1, FC=3, Token=0000000B, length=11:

Temp 21.464828

 

HandleData 12 10 11 12

[N] 2472| main.cpp             [  50] handleData: Response: serverID=1, FC=3, Token=0000000C, length=11:

Level -0.000473

 

With the LOG_N line commented out, here is the output:

 

mode:DIO, clock div:2

load:0x3fff0030,len:1184

load:0x40078000,len:13232

load:0x40080400,len:3028

entry 0x400805e4

__ OK __

 

Count 0

Wakeup 

 

Count 1        

 

Pressure req 10

Temp req 11    

Level req 12   

 

HandleData 10 10 11 12

Pressure -0.001928    

 

HandleData 12 10 11 12

Level 0.000213        

 

It repeats this sequence of missing sending the temp reading with about a 2 second delay for the Level reading. 

 

Here are scope plots of the good and the bad. Note the very different timebases.

 

Regards,
Ron

 

Good



 

Bad:

**Reply by Miq1** [reply]
I tried to compact your responses a bit, but still the pictures are missing etc.

The LOG_N line is taking some time - may be that makes the difference. If your server is responding really fast, the callbacks will be fired without much time in between (mind, the callbacks are called in the client task, independently of the main loop() task). So you may get into a race condition between the Serial outputs of the main and client tasks. I seem to remember the Serial calls are sequentialized by the core, but may be wrong here.

Just an idea, nothing I could prove.

**Reply by Miq1** [reply]
Wait a second! You wrote:
> It repeats this sequence of missing sending the temp reading with about a 2 second delay for the Level reading.

This sounds exactly as if your server was not able to respond that fast. Normally an E0 TIMEOUT error should have been responded, but will not be seen due to your LOG_LEVEL being set to 0.

I would suggest to raise the quiet time between requests a bit or for trying out simply put some delay() between requests. Alternatively, change the LOG_E in your error handler into a Serial.printf to make sure it will be seen.

**Comment by ghost** [comment]
It works fine without any logging macros in the handleData callback if I insert a delay of at least 150mSec between the three requests. That puts the next request 20 or so mSecs past the previous reply processing (see below). Perhaps the problem is something to do with the serialisation. I assume I don’t need to worry about the number of queues, etc.

 

The total response time is still well under my criterion of 1 second, so that will have to do.

 

Regards,
Ron

**Reply by Miq1** [reply]
Using ``delay()`` is no real solution to my liking... What is happening in principle:
- ``addRequest()`` puts a request on the queue
- with the next idle loop the client task picks it from the queue and sends it
- while the response is expected and we are still within the set timeout time, further ``addRequest()`` calls may put more requests on the queue
- When either a response is received or the timeout strikes, the client will call your respective callback with it
- When the callback returns, the next request is picked and sent, but not before the minimum quiet time has passed (~1,7ms with higher baud rates)

Since this is all done in the independent client task, whatever you do in ``loop()`` will have no direct influence on it.

So why does a ``delay()`` help now? The client will remain idle after the last response has been processed for the delay time minus the time for sending the last request, waiting for and processing the response. Sending and processing is a matter of a few milliseconds, so the variable part is the time the server takes to respond.

If the server is responding slowly, the idle time may be (almost) used up already before the response is received and processed. Thus the next request ist sent immediately after the bus quiet time after the response has passed, leaving less time for the server to recover. If the idle time is prolonged by your ``delay()`` that gives the server more time to recover.

Concluding I would guess the server is driven to and even over its timing limits by the three consecutive requests.

My solution proposal would be to do a single request for 20 registers starting at 0x25 and to pick just the wanted registers from the response in the data handler. Hence the server would have to answer a single request every 3 seconds only and would get plenty of time to catch up.

**Comment by ghost** [comment, marked as answer]
Hi Michael

I am sure you are correct; the three consecutive requests cause some internal problem in the library. Your final suggestion was the solution – I read the entire sensor data space (51 bytes) and in the handleData callback, copy the response data to another buffer which I unpack elsewhere with no time pressure. And no logging code required.

Many thanks,

Ron

**Reply by Miq1** [reply]
Ahem... No internal problem in the library :wink: I have a sample sketch firing 1000 requests at 5 Mbaud without a pause and without any issue, because the server is responding equally fast.

But good you got it working as you need it.

**Comment by ghost** [comment]
Sorry. I misunderstood you. But yes, all working as expected now.

Regards,
Ron
